ITISFoundation/osparc-simcore
# ✨ Users pre-registration  🗃️ 🚨

## What do these changes do?

This PR introduces a new step in the workflow to register an account denoted "pre-registration" and is performed by the PO upon approval of the request. This is how it goes:
1. User **requests an account** by filling an online form 
    - Form and geo-location is sent to PO for approval
    - Some of the info is legaly required to produce invoices (e.g. billing address etc)
2. When request is **approved**, then the user is **pre-registered** (PO center)
    - Part of the information in the request form is added as a pre-registration
3. The user receives an invitation link  (PO center)
    - The invitation is still manually sent by the PO but when https://github.com/ITISFoundation/osparc-simcore/pull/5333 is merged we could send the email automatically as well.
4. When the invited user **registers**, the new account is merged with pre-registration details already in the system
    - During registration, this information will be used to setup wallet names, descriptions and welcome message (i.e. profile) as requested by tests users.
    - The billing information necessary to produce invoices as well (see `test_user_creation_workflow` )


### Highlights
-  appends `ipinfo` for geo-location on "request account" form
- ✨ NEW API for pre-registration `webserver v0.38.0`
    - `GET /v0/users:search` (`search_users`) to search users by email
    - `POST /v0/users:pre-register` (`pre_register_user`) to pre-register  a user 
- Use extra info to setup wallet names and welcome messages
- Pre-registration injects data in `users_details` and is linked with `users` upon registration
    - 🗃️ New `users_details` table
- ♻️ `webserver.users` plugin is split up
- **Follow up**: 
    - will build a GUI for the PO center for the pre-registration (with @elisabettai @odeimaiz )
    - will connect with https://github.com/ITISFoundation/osparc-simcore/pull/5402 (with @matusdrobuliak66 )

## Related issue/s

- Implements pre-registration in #5138
- Resolves https://github.com/ITISFoundation/osparc-issues/issues/1193


## How to test

#### Driving tests
```
cd services/web/server
make install-dev
pytest -vv -k test_pre_registration_and_invitation_workflow tests/unit
```
```
cd packages/postgres-database
make install-dev
pytest -vv tests/test_users_details.py
```
#### Manual testing 🚨

- Fill and send request form
  - PO should receive full form and ip info 
- invite a new user (w/o pre-registration)  
  - should work as before 
  - check welcome during registration
- invite a new user w/ pre-registration
  - pre-register via swagger site (GUI still not available)
  - check welcome during registration

## Dev Checklist


## DevOps Checklist

- 🗃️ new `invited_users` table
